### PR TITLE
Render activity list items fully first when they enter viewport

### LINKED ISSUE #2396: Activity archive loads everything at once (and prevents further navigation)
## Description
Activities (events, call assignments etc) end up in the archive after they have taken place. This means that the archive can become a very long list of activities. The list items contain stats that need to be loaded per activity, so when the list is long that results in a very large number of HTTP requests being queued up in the browser. Besides taking a long time to load, it also means that the user can't navigate further, because the async loading required to render the next "page" will be queued after all of the stats requests.

I think the solution to this is to only load the stats for activities that are currently visible.

## Steps to reproduce
1. Go to https://app.dev.zetkin.org/organize/1/projects
2. Click on "Archive" or "Activities" (whichever is longer)
3. Toggle visibility on for all types of activities, to extend the length of the list
4. Wait for everything to load
5. Scroll to bottom of list

### Expected Behaviour
In step 4, only the visible activities should load their stats, and when you scroll down, some more should load.

### Actual Behaviour
All activities load fully in step 4, so when you scroll down in step 5 nothing additional happens.

## Screenshots (if you have any)
image

### PR DESCRIPTION
**Pull request by awarn**
## Description

Wraps the items of activity list in an IntersectionObserver component that waits for them to be inside the viewport before rendering them (which defers loading until then as well).

Displays a 'pending' list item to look a little neater on loading flash, and also take about roughly the same space as the actual items would.

## Related issues
Resolves #2396

### CODE REVIEW

**Comment by awarn** [comment]
I should really clean this up a little before merge, the new component can be its own file. And also we maybe should discuss an alternative: Extending the actual list items (CallAssignmentListItem, etc) with logic to wait with loading their data but render their own 'pending' state rather than relying on stopping their render as I did here. My current version was just quicker than modifying them all. Tell me what you prefer, I see some pros and cons with both.

**Comment by richardolsson** [comment]
This is an impressive solution! There have been plans to make some major changes to the pages where these lists live, so I don't think it is necessary to make this perfect. I'll review properly, but just wanted to stop in real quick and let you know that you should feel no pressure to clean this up!

File: src/features/campaigns/components/ActivityList/index.tsx [hunk 1]
```diff
@@ -17,12 +17,61 @@ import useClusteredActivities, {
...
+}: LazyActivitiesBoxProps) => {
+  const [inView, setInView] = useState(false);
+  const boxRef = useRef<HTMLElement>();
+
+  useEffect(() => {
+    if (!boxRef.current) {
+      return;
+    }
+
+    const observer = new IntersectionObserver((entries) => {
```
Comment (richardolsson) [inline src/features/campaigns/components/ActivityList/index.tsx:44]: I'm so glad you introduced this solution. I was not aware of `IntersectionObserver` (like I was with `ResizeObserver` and observers in general) so I had to look it up, but it's covered by all browsers that we care about so I think it's the ultimate solution for us.

File: src/features/campaigns/components/ActivityList/index.tsx [hunk 2]
```diff
@@ -17,12 +17,61 @@ import useClusteredActivities, {
...
+      entries.forEach((entry) => {
+        if (entry.isIntersecting) {
+          setInView(true);
+        }
+      });
+    });
+
+    observer.observe(boxRef.current);
```
Comment (richardolsson) [inline src/features/campaigns/components/ActivityList/index.tsx:52]: Do we need to unobserve or destroy the observer when the component unmounts, or will that be handled automatically by the browser when React removes the element from the DOM?

**Review by richardolsson — APPROVED** [review]
I have reviewed the code and I've tried the build. I confirmed that not all data is loaded on first draw, but more data is loaded as the user scrolls.

I'm going to approve it as is because I'm eager to get this merged. But please note my question about garbage collection below and if you believe that something needs to be done let me know and we can create a separate issue for that. 😊

File: src/features/campaigns/components/ActivityList/index.tsx (hunk 2 above)
Comment (awarn) [reply src/features/campaigns/components/ActivityList/index.tsx:52]: From what I can find in discussions the intent is for garbage collection to deal with it if observer is no longer observing any elements (using weak references):
[https://github.com/w3c/csswg-drafts/issues/6397](https://github.com/w3c/csswg-drafts/issues/6397)

**Pull request merged**